src/ray/id.h

```diff
@@ -39,7 +64,59 @@ class RAY_EXPORT UniqueID {
 
 static_assert(std::is_standard_layout<UniqueID>::value, "UniqueID must be standard");
 
+class TaskID : public BaseID<TaskID> {
+ public:
+  TaskID() : BaseID() {}
+  size_t hash() const;
+  static size_t size() { return kUniqueIDSize - sizeof(int32_t); }
+  static TaskID GetDriverTaskID(const DriverID &driver_id);
+
+ protected:
+  uint8_t id_[kUniqueIDSize - sizeof(int32_t)];
+};
+
+class ObjectID : public BaseID<ObjectID> {
+ public:
+  ObjectID() : BaseID() {}
+  size_t hash() const;
+  static size_t size() { return kUniqueIDSize; }
+  plasma::ObjectID to_plasma_id() const;
+  ObjectID(const plasma::UniqueID &from);
+
+  /// Get the index of this object in the task that created it.
+  ///
+  /// \return The index of object creation according to the task that created
+  /// this object. This is positive if the task returned the object and negative
+  /// if created by a put.
+  int32_t object_index() const { return index_; }
+
+  /// Compute the task ID of the task that created the object.
+  ///
+  /// \return The task ID of the task that created this object.
+  const TaskID &task_id() const { return task_id_; }
+
+  /// Compute the object ID of an object put or returned by the task.
+  ///
+  /// \param task_id The task ID of the task that created the object.
+  /// \param index What number the object was created by in the task.
+  /// \return The computed object ID.
+  static ObjectID build(const TaskID &task_id, bool is_put, int64_t index);
+
+ protected:
+  TaskID task_id_;
+  int32_t index_;
+  mutable size_t hash_ = 0;
```

Comment:
why not define `hash_` in `BaseID`?

Reply:
The same as the TaskID. If we will nest some IDs into another ID, we cannot use cache_. Otherwise, we need to do some copy past operation in binary().